### PR TITLE
[1.16.x] Add support for forge's "remove" list to tag datagenerators

### PR DESCRIPTION
**Pull request by Commoble**
## Background

Forge expands the tag format to support a "remove" list, allowing tags to specify IDs of individual elements or IDs of tags to remove from that tag rather than adding to it. However, the current tag generators do not support this; tag generators cannot currently generate tag jsons that use the "remove" list. This PR expands the tag datagenerators to support this.

For clarification, there are two kinds of "tag builders" that this PR touches:
* Firstly is the ITag.Builder (sometimes called the "internal" or "raw" builder, which is used for both parsing tag jsons when data is loaded, and for writing tag jsons during datagen. This builder only works with the IDs of things and doesn't care about what the tags actually represent.
* The other one is the TagsProvider.Builder, which is exclusively used for datagenerating tags of registry objects. TagsProviders provide TagsProviders.Builders to callers; these are wrappers around ITag.Builders but with methods for making tags of registry objects.

## What's this PR do?
We patch ITag.Builder in the following ways:
* We implement a new IForgeRawTagBuilder forge extension interface to minimize the size of the patch
* We add a list field to ITag.Builder that items to add to the "remove" list in the json can be added to
* Where the builder serializes its data to a json, we add a line to serialize our extra removal data

IForgeRawTagBuilder has methods for adding to the removal list in a similar manner to the way ITag.Builder adds values to the value list.

We also add extra methods to IForgeTagBuilder (the extension interface of TagsProvider.Builder) to assist with adding registry objects and tags to the removal list in the same way it can add things to the standard values list.

The reason for adding these methods to both builders is that A) data generation of tags of registry objects should use TagsProviders and their builders rather than working with ITag.Builders directly, while B) tags of things that aren't registry objects can easily exist (see: function tags) and would necessarily need their own data providers (not TagsProviders) backed by ITag.Builders in a similar way.

### CODE REVIEW

**Comment by ChampionAsh5357** [comment]
As the 1.17 version of the PR has been merged, the changes should be backported to this version. Apologies for not getting around to reviewing this until now.

File: patches/minecraft/net/minecraft/tags/ITag.java.patch [hunk 1]
```diff
@@ -1,10 +1,15 @@
...
+       return Tag.func_241286_a_(p_232946_0_);
+    }
  
-    public static class Builder {
+-   public static class Builder {
++   public static class Builder implements net.minecraftforge.common.extensions.IForgeRawTagBuilder {
        private final List<ITag.Proxy> field_232953_a_ = Lists.newArrayList();
 +      private boolean replace = false;
++      private final List<ITag.Proxy> removeEntries = Lists.newArrayList();
++      public List<ITag.Proxy> getRemoveEntries() { return this.removeEntries; } // forge -- track additions to the "remove" list during datageneration
```
Comment (ChampionAsh5357) [inline patches/minecraft/net/minecraft/tags/ITag.java.patch:12]: This should be a read-only view, see [this](https://github.com/MinecraftForge/MinecraftForge/pull/7911/files#diff-f8049b7220c9c3128f74e20f2128209a7a56cc09f4306035144cb8ed4dfbfdb7R9-R15).

File: src/main/java/net/minecraftforge/common/extensions/IForgeRawTagBuilder.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        }
+    }
+    
+    /**
+     * Adds a tag proxy to the remove list to generate a json with
+     * @param proxy Tag proxy (either an item or another tag, usually)
+     * @return builder for chaining
+     * @apiNote internal; the other methods are preferable to call
+     */
+    default ITag.Builder removeProxy(final ITag.Proxy proxy)
```
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/common/extensions/IForgeRawTagBuilder.java:61]: See [Lex's comment](https://github.com/MinecraftForge/MinecraftForge/pull/7911/files#r684373367) on this.

File: src/main/java/net/minecraftforge/common/extensions/IForgeRawTagBuilder.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        return rawBuilder;
+    }
+    
+    /**
+     * Adds a tag entry to the remove list.
+     * @param entry The tag entry to add to the remove list
+     * @param source The source of the caller for logging purposes (generally a modid)
+     * @return The builder for chaining purposes
+     */
+    default ITag.Builder removeTagEntry(final ITag.ITagEntry tagEntry, final String source)
```
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/common/extensions/IForgeRawTagBuilder.java:74]: Use same naming scheme as in 1.17 to reduce upgrading or downgrading complexity.

File: src/main/java/net/minecraftforge/common/extensions/IForgeRawTagBuilder.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        return this.removeProxy(new ITag.Proxy(tagEntry,source));
+    }
+    
+    /**
+     * Adds a single-element entry to the remove list.
+     * @param elementID The ID of the element to add to the remove list
+     * @param source The source of the caller for logging purposes (generally a modid)
+     * @return The builder for chaining purposes
+     */
+    default ITag.Builder removeElementByID(final ResourceLocation elementID, final String source)
```
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/common/extensions/IForgeRawTagBuilder.java:85]: Use same naming scheme as in 1.17 to reduce upgrading or downgrading complexity.

File: src/main/java/net/minecraftforge/common/extensions/IForgeRawTagBuilder.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    
+    /**
+     * Adds a tag to the remove list.
+     * @param tagID The ID of the tag to add to the remove list
+     * @param source The source of the caller for logging purposes (generally a modid)
+     * @return The builder for chaining purposes
+     */
+    default ITag.Builder removeTagByID(final ResourceLocation tagID, final String source)
```
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/common/extensions/IForgeRawTagBuilder.java:97]: Use same naming scheme as in 1.17 to reduce upgrading or downgrading complexity.

File: src/main/java/net/minecraftforge/common/extensions/IForgeTagBuilder.java [hunk 6]
```diff
@@ -67,4 +67,82 @@
     {
         return getBuilder().add(new ITag.OptionalTagEntry(location));
     }
+    
+    /**
+     * Adds a registry entry to the tag json's remove list. Callable during datageneration.
+     * @param element The entry to remove
+     * @return The builder for chaining
+     */
+    default TagsProvider.Builder<T> removeRegistryEntry(final IForgeRegistryEntry<T> entry)
```
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/common/extensions/IForgeTagBuilder.java:76]: Use same naming scheme as in 1.17 to reduce upgrading or downgrading complexity.

File: src/main/java/net/minecraftforge/common/extensions/IForgeTagBuilder.java [hunk 7]
```diff
@@ -67,4 +67,82 @@
...
+        return removeElementByID(entry.getRegistryName());
+    }
+    
+    /**
+     * Adds multiple registry entries to the tag json's remove list. Callable during datageneration.
+     * @param entries The entries to remove
+     * @return The builder for chaining
+     */
+    @SuppressWarnings("unchecked")
+    default TagsProvider.Builder<T> removeRegistryEntries(final IForgeRegistryEntry<T>...entries)
```
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/common/extensions/IForgeTagBuilder.java:87]: Use same naming scheme as in 1.17 to reduce upgrading or downgrading complexity.

File: src/main/java/net/minecraftforge/common/extensions/IForgeTagBuilder.java [hunk 8]
```diff
@@ -67,4 +67,82 @@
...
+        }
+        return getBuilder();
+    }
+    
+    /**
+     * Adds a single element's ID to the tag json's remove list. Callable during datageneration.
+     * @param location The ID of the element to remove
+     * @return The builder for chaining
+     */
+    default TagsProvider.Builder<T> removeElementByID(final ResourceLocation location)
```
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/common/extensions/IForgeTagBuilder.java:101]: Use same naming scheme as in 1.17 to reduce upgrading or downgrading complexity.

File: src/main/java/net/minecraftforge/common/extensions/IForgeTagBuilder.java [hunk 9]
```diff
@@ -67,4 +67,82 @@
...
+        builder.getInternalBuilder().removeElementByID(location, builder.getModID());
+        return builder;
+    }
+    
+    /**
+     * Adds multiple elements' IDs to the tag json's remove list. Callable during datageneration.
+     * @param locations The IDs of the elements to remove
+     * @return The builder for chaining
+     */
+    default TagsProvider.Builder<T> removeElementsByID(final ResourceLocation... locations)
```
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/common/extensions/IForgeTagBuilder.java:113]: Use same naming scheme as in 1.17 to reduce upgrading or downgrading complexity.

File: src/main/java/net/minecraftforge/common/extensions/IForgeTagBuilder.java [hunk 10]
```diff
@@ -67,4 +67,82 @@
...
+        }
+        return getBuilder();
+    }
+    
+    /**
+     * Adds a tag to the tag json's remove list. Callable during datageneration.
+     * @param tag The ID of the tag to remove
+     * @return The builder for chaining
+     */
+    default TagsProvider.Builder<T> removeTag(ITag.INamedTag<T> tag)
```
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/common/extensions/IForgeTagBuilder.java:127]: Use same naming scheme as in 1.17 to reduce upgrading or downgrading complexity.

File: src/main/java/net/minecraftforge/common/extensions/IForgeTagBuilder.java [hunk 11]
```diff
@@ -67,4 +67,82 @@
...
+        return builder;
+    }
+    
+    /**
+     * Adds multiple tags to the tag json's remove list. Callable during datageneration.
+     * @param tags The IDs of the tags to remove
+     * @return The builder for chaining
+     */
+    @SuppressWarnings("unchecked")
+    default TagsProvider.Builder<T> removeTags(ITag.INamedTag<T>...tags)
```
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/common/extensions/IForgeTagBuilder.java:140]: Use same naming scheme as in 1.17 to reduce upgrading or downgrading complexity.

**Review by ChampionAsh5357 — CHANGES_REQUESTED** [review]
Mainly just missing functionalities and consistent namings between the two PRs.

**Comment by sciwhiz12** [comment]
Hello! @Commoble, please leave a comment whether you wish to continue this PR and update it to the latest active version of 1.18 (or leave an explanation why this PR's changes are not applicable to 1.18). If we do not hear back from you after some time, this PR will be closed for inactivity.
If anyone else wishes to continue the work of this PR, please create a PR with a link to this PR for reference. 
Thank you for your contributions. 👋

**Comment by Commoble** [comment]
This was merged into 1.18 in #7911; I have no further interest in backporting this at this time, but anyone else who needs this in 1.16 is welcome to

**Pull request closed without merging**